File: stacks/python/base/code/cache.tpl.py
"""
File: cache.tpl.py
Purpose: Redis caching utilities with async support
Generated for: {{PROJECT_NAME}}
"""

import json
import hashlib
from typing import Any, Callable, Optional, TypeVar, Union
from functools import wraps
from datetime import timedelta
import redis.asyncio as redis
from pydantic import BaseModel

T = TypeVar("T")


class CacheConfig(BaseModel):
    host: str = "localhost"
    port: int = 6379
    db: int = 0
    password: Optional[str] = None
    key_prefix: str = "app:"
    default_ttl: int = 3600  # 1 hour


class RedisCache:
    def __init__(self, config: CacheConfig):
        self.config = config
        self.client: Optional[redis.Redis] = None

    async def connect(self):
        self.client = redis.Redis(
            host=self.config.host,
            port=self.config.port,
            db=self.config.db,
            password=self.config.password,
            decode_responses=True,
        )

    async def disconnect(self):
        if self.client:
            await self.client.close()

    def _key(self, key: str) -> str:
        return f"{self.config.key_prefix}{key}"

    async def get(self, key: str) -> Optional[Any]:
        """Get a value from cache"""
        value = await self.client.get(self._key(key))
        if value:
            return json.loads(value)
        return None

    async def set(
        self,
        key: str,
        value: Any,
        ttl: Optional[int] = None,
    ) -> None:
        """Set a value in cache"""
        ttl = ttl or self.config.default_ttl
        await self.client.setex(
            self._key(key),
            ttl,
            json.dumps(value, default=str),
        )

    async def delete(self, key: str) -> None:
        """Delete a key from cache"""
        await self.client.delete(self._key(key))

    async def delete_pattern(self, pattern: str) -> int:
        """Delete all keys matching a pattern"""
        keys = await self.client.keys(self._key(pattern))
        if keys:
            return await self.client.delete(*keys)
        return 0

    async def exists(self, key: str) -> bool:
        """Check if a key exists"""
        return await self.client.exists(self._key(key)) > 0

    async def incr(self, key: str, amount: int = 1) -> int:
        """Increment a counter"""
        return await self.client.incrby(self._key(key), amount)

    async def expire(self, key: str, ttl: int) -> None:
        """Set TTL on existing key"""
        await self.client.expire(self._key(key), ttl)


# Decorator for caching function results
def cached(
    cache: RedisCache,
    key_prefix: str,
    ttl: int = 3600,
    key_builder: Optional[Callable[..., str]] = None,
):
    """Decorator to cache function results"""
    def decorator(func: Callable[..., T]) -> Callable[..., T]:
        @wraps(func)
        async def wrapper(*args, **kwargs) -> T:
            # Build cache key
            if key_builder:
                cache_key = key_builder(*args, **kwargs)
            else:
                key_data = json.dumps({"args": args, "kwargs": kwargs}, sort_keys=True, default=str)
                key_hash = hashlib.md5(key_data.encode()).hexdigest()[:12]
                cache_key = f"{key_prefix}:{key_hash}"

            # Try to get from cache
            cached_value = await cache.get(cache_key)
            if cached_value is not None:
                return cached_value

            # Execute function and cache result
            result = await func(*args, **kwargs)
            await cache.set(cache_key, result, ttl)
            return result

        return wrapper
    return decorator


# Global cache instance (configure at startup)
cache: Optional[RedisCache] = None


async def init_cache(config: CacheConfig) -> RedisCache:
    global cache
    cache = RedisCache(config)
    await cache.connect()
    return cache


# Usage:
# cache = await init_cache(CacheConfig())
# 
# @cached(cache, "users", ttl=300)
# async def get_user(user_id: str):
#     return await db.fetch_user(user_id)

File: stacks/node/base/code/cache.tpl.js
/*
File: cache.tpl.js
Purpose: Redis caching utilities with ioredis
Generated for: {{PROJECT_NAME}}
*/

const Redis = require('ioredis');
const crypto = require('crypto');

class RedisCache {
    constructor(config = {}) {
        this.config = {
            host: config.host || 'localhost',
            port: config.port || 6379,
            db: config.db || 0,
            password: config.password,
            keyPrefix: config.keyPrefix || 'app:',
            defaultTTL: config.defaultTTL || 3600,
        };

        this.client = new Redis({
            host: this.config.host,
            port: this.config.port,
            db: this.config.db,
            password: this.config.password,
            retryStrategy: (times) => Math.min(times * 50, 2000),
        });
    }

    _key(key) {
        return `${this.config.keyPrefix}${key}`;
    }

    async get(key) {
        const value = await this.client.get(this._key(key));
        if (value) {
            try {
                return JSON.parse(value);
            } catch {
                return value;
            }
        }
        return null;
    }

    async set(key, value, ttl = null) {
        const serialized = JSON.stringify(value);
        const expiry = ttl || this.config.defaultTTL;
        await this.client.setex(this._key(key), expiry, serialized);
    }

    async delete(key) {
        await this.client.del(this._key(key));
    }

    async deletePattern(pattern) {
        const keys = await this.client.keys(this._key(pattern));
        if (keys.length > 0) {
            await this.client.del(...keys);
        }
        return keys.length;
    }

    async exists(key) {
        return (await this.client.exists(this._key(key))) > 0;
    }

    async incr(key, amount = 1) {
        return this.client.incrby(this._key(key), amount);
    }

    async expire(key, ttl) {
        await this.client.expire(this._key(key), ttl);
    }

    async getOrSet(key, factory, ttl = null) {
        const cached = await this.get(key);
        if (cached !== null) {
            return cached;
        }

        const value = await factory();
        await this.set(key, value, ttl);
        return value;
    }

    async disconnect() {
        await this.client.quit();
    }
}

// Cache decorator for functions
function cached(cache, keyPrefix, ttl = 3600, keyBuilder = null) {
    return function (target, propertyKey, descriptor) {
        const originalMethod = descriptor.value;

        descriptor.value = async function (...args) {
            let cacheKey;
            if (keyBuilder) {
                cacheKey = keyBuilder(...args);
            } else {
                const keyData = JSON.stringify(args);
                const hash = crypto.createHash('md5').update(keyData).digest('hex').slice(0, 12);
                cacheKey = `${keyPrefix}:${hash}`;
            }

            const cachedValue = await cache.get(cacheKey);
            if (cachedValue !== null) {
                return cachedValue;
            }

            const result = await originalMethod.apply(this, args);
            await cache.set(cacheKey, result, ttl);
            return result;
        };

        return descriptor;
    };
}

// Functional wrapper for caching
function withCache(cache, keyPrefix, ttl = 3600) {
    return function (fn) {
        return async function (...args) {
            const keyData = JSON.stringify(args);
            const hash = crypto.createHash('md5').update(keyData).digest('hex').slice(0, 12);
            const cacheKey = `${keyPrefix}:${hash}`;

            const cachedValue = await cache.get(cacheKey);
            if (cachedValue !== null) {
                return cachedValue;
            }

            const result = await fn(...args);
            await cache.set(cacheKey, result, ttl);
            return result;
        };
    };
}

module.exports = {
    RedisCache,
    cached,
    withCache,
};

// Usage:
// const cache = new RedisCache({ keyPrefix: 'myapp:' });
// const getUser = withCache(cache, 'users', 300)(async (id) => {
//   return await db.findUser(id);
// });
